federationexample/README.md

```diff
@@ -0,0 +1,135 @@
+
+## Architecture Diagram
+
+![Screenshot](federation.png)
+
+## API 
+
+If an object entirely belongs to a gqlserver, it can exist on only that gqlserver with all it's field funcs. However, there are some objects that have fields that belong to different gqlservers. We call these federated objects. 
+
+A nonfederated object looks something like this example below 
+```
+type Device struct {
+    Id            int64
+    OrgId         int64
+    Name          string
+}
+device := schema.Object("Device", Device{})
+```
+
+A federated object has an extra field called FetchObjectFromKeys that takes a function, telling the gqlserver how to fecth the full object from a the federated keys, the minimum number of fields needed to fetch the object. A federated object can exist on multiple gqlservers and hhave different field funcs on each of the team's gqlservers. 
+
+```
+type Device struct {
+    Id            int64
+    OrgId         int64
+    Name          string
+}
+// Device keys has to be a subset of the fields on Device and is the minimum information needded to fetch a full Device
+type DeviceKeys struct {
+    Id            int64
+    OrgId         int64
+}
+device := schema.Object("Device", Device{}, schemabuilder.FetchObjectFromKeys(func(args struct{ Keys []*DeviceKeys }) []*Device {
+    // Fetch device from database using device keys    
+    return fetchedDevices
+}))
+```
+
+
+## Example
+
+In teh federation example, we have 2 gqlservers (device and safety). Device and Location are federated objects, whcih allows them to have field funcs on both gqlservers. 
+```
+{
+  device {                  #devicegqlserver
+    id                      #devicegqlserver
+    location {              #devicegqlserver
+      latitude              #devicegqlserver
+      longitude             #devicegqlserver
+      street                #devicegqlserver
+    }
+    camera{                 #safetygqlserver
+      id                    #safetygqlserver
+      isOn                  #safetygqlserver
+      isMultiCam            #safetygqlserver
+      cameraLocation {      #safetygqlserver
+      	latitude            #safetygqlserver
+      	longitude           #safetygqlserver
+        street              #devicegqlserver
+   	  }
+    }
+  }
+}
+```
+
+
+When the gateway parses this query, it first makes a request to devicegqlserver. It adds all the fields that exist on device server, and when it sees a field it doesnt know how to resolve it replalces it with the field func `__federation` and adds theh deviceKeys are fields nested on `__federation`.
+```
+{
+  device {                  #devicegqlserver
+    id                      #devicegqlserver
+    location {              #devicegqlserver
+      latitude              #devicegqlserver
+      longitude             #devicegqlserver
+      street                #devicegqlserver
+    }
+    _federation {
+        id
+        orgId
+    }
```

Comment:
This bit is also confusing.  I assume a few things
 - `_federation` -> `__federation` per above description?
 - Which `id` and `orgId` are there referring to from the original query in this example (camera.id and camera.orgId, respectively?)

Reply:
* We had to change it to one underscore to match the apollo schema validation. The double underscore is reserved for fields used for introspection
* This would be referring to the id and org id on the device. I also added this to the description above